core/src/main/kotlin/net/corda/core/transactions/TransactionWithSignatures.kt

```diff
@@ -103,6 +106,48 @@ interface TransactionWithSignatures : NamedByHash {
         val sigKeys = sigs.mapToSet { it.by }
         // TODO Problem is that we can get single PublicKey wrapped as CompositeKey in allowedToBeMissing/mustSign
         //  equals on CompositeKey won't catch this case (do we want to single PublicKey be equal to the same key wrapped in CompositeKey with threshold 1?)
-        return requiredSigningKeys.asSequence().filter { !it.isFulfilledBy(sigKeys) }.toSet()
+        return requiredSigningKeys.asSequence().filter { requiredSigningKey -> !isKeyInSignature(requiredSigningKey, sigKeys) }.toSet()
+    }
+
+    private fun isKeyInSignature(requiredSigningKey: PublicKey, signingKeys: Set<PublicKey>): Boolean {
+        // This bit should be after requiredSigningKey.isFulfilledBy(signingKey) check
+        // The flag validatedByKeyRotationCheck is not required. this is only for testing purposes
+        ///
+        val oldPublicKeyToCrossProviderKeyRotationProofMap =
+                sigs.map { signature ->
+                    signature.signatureMetadata.crossProviderKeyRotationProof.map { keyRotationProof ->
+                        keyRotationProof.publicKeyOld to keyRotationProof
+                    }
+                }.flatten().toMap()
+
+        var validatedByKeyRotationCheck = false
+        var oldPublicKeyToKeyRotationState = oldPublicKeyToCrossProviderKeyRotationProofMap[requiredSigningKey]
+        while(oldPublicKeyToKeyRotationState != null)  {
+            validatedByKeyRotationCheck = true
+            loggerFor<TransactionSignature>().info("Filipe: Validating old key. Old key: ${oldPublicKeyToKeyRotationState.publicKeyOld}")
+            val isKeyRotationProofValid = doVerify(
+                    findSignatureScheme(requiredSigningKey),
+                    oldPublicKeyToKeyRotationState.publicKeyOld,
+                    oldPublicKeyToKeyRotationState.signature,
+                    oldPublicKeyToKeyRotationState.publicKeyNew.encoded
+            )
+            if(!isKeyRotationProofValid) {
+                loggerFor<TransactionSignature>().info("Filipe: Old key is invalid")
+                return false
+            }
+
+            oldPublicKeyToKeyRotationState = oldPublicKeyToCrossProviderKeyRotationProofMap[oldPublicKeyToKeyRotationState.publicKeyNew]
+        }
+
+        if(validatedByKeyRotationCheck) {
+            return true
+        }
+        ////
+
+
+        if(requiredSigningKey.isFulfilledBy(signingKeys)) {
+            return true
+        }
+        return false
```

Comment:
I think this would make it working for composite keys. (assuming newPublicKeyToOldMap has the verified relationship of keys extracted from the signatures.)

```suggestion
        val allSigningKeys = sigKeys + sigkeys.mapNotNull { newPublicKeyToOldMap[it]}
        return requiredSigningKeys.asSequence().filter { !it.isFulfilledBy(allSigningKeys) }.toSet()
    }
```

Reply:
Thanks, @vlajos. Your approach is much simpler and works just as well. I have updated the code based on your suggestion to support composite keys.